src/services/firm_quote_validator.ts

```diff
@@ -0,0 +1,147 @@
+        if (makerTokenAddresses.length !== 1) {
+            logger.error(`Quotes array was empty or found multiple maker token addresses within one single RFQ batch: ${JSON.stringify(makerTokenAddresses)}. Rejecting the batch`);
+            return quotes.map(_quote => BIG_NUMBER_ZERO);
+        }
+        const makerTokenAddress = makerTokenAddresses[0];
+
+        // Fetch balances and create a lookup table
+        const makerLookup: {[key: string]: BigNumber} = {};
+        const makerAddresses = Array.from(new Set(quotes.map(quote => quote.makerAddress)));
+        const cacheResults = await this._chainCacheRepository.find({
+            where: [{
+                tokenAddress: makerTokenAddresses[0],
+                makerAddress: In(makerAddresses),
+            }],
+        });
+        const nowUnix = (new Date()).getTime();
+        for (const result of cacheResults) {
+            makerLookup[result.makerAddress!] = this._calculateMakerBalanceFromResult(result, makerTokenAddress, nowUnix);
+        }
+
+        // Finally, adjust takerFillableAmount based on maker balances
+        const {makerAddressesToAddToCache, takerFillableAmounts} = PostgresBackedFirmQuoteValidator._calculateTakerFillableAmountsFromQuotes(quotes, makerLookup);
+
+        // If any new addresses were found, add new addresses to cache.
+        // NOTE: since this insertion happens on the web processes, we need to gracefully handle conflict
+        // that can happen if two threads try to insert the same entry at the same time. This is why we add
+        // the "ON CONFLICT" clause.
+        if (makerAddressesToAddToCache.length > 0) {
+            logger.info(`Adding new addresses to cache: ${JSON.stringify(makerAddressesToAddToCache)}`);
+            await this._chainCacheRepository
+                .createQueryBuilder()
+                .insert()
+                .values(
+                    makerAddressesToAddToCache.map(makerAddress => {
+                        return {
+                            makerAddress,
+                            tokenAddress: makerTokenAddresses[0],
+                            timeFirstSeen: 'NOW()',
+                        };
+                    })
+                )
+                .onConflict(`("token_address", "maker_address") DO NOTHING`)
+                .execute();
+        }
+        return takerFillableAmounts;
+    }
+
+    private _calculateMakerBalanceFromResult(result: MakerBalanceChainCacheEntity, makerTokenAddress: string, nowUnix: number): BigNumber {
+        if (!result.timeOfSample) {
+            // If a record exists but a time of sample does not yet exist, this means that the cache entry has not yet been
+            // populated by the worker process. This may be due to a new address being added a few minutes ago, but it could
+            // also be due to a bug in the worker.
+            const timeFirstSeen = result.timeFirstSeen ? result.timeFirstSeen.getTime() : 0;
```

Comment:
is this in ms?

Reply:
yes it is, do you think I should specify in the naming (`timeOfFirstSeenMs`?)